File: src/dsl/evaluator.js
import { defaultContext } from './builtins.js';

function evalNode(node, ctx) {
  switch (node.type) {
    case 'NumberLiteral':
      return node.value;
    case 'StringLiteral':
      return node.value;
    case 'BooleanLiteral':
      return node.value;
    case 'NullLiteral':
      return null;
    case 'Identifier': {
      const name = node.name;
      if (ctx.operations[name]) return ctx.operations[name];
      if (ctx.surfaces[name]) return ctx.surfaces[name];
      if (ctx.enums && ctx.enums[name]) return ctx.enums[name];
      return name;
    }
    case 'MemberExpr': {
      const prop = node.property.name;
      if (node.object.type === 'Identifier' && ctx.enumMethods) {
        const objName = node.object.name;
        const methods = ctx.enumMethods[objName];
        if (methods && methods[prop]) {
          return methods[prop];
        }
      }
      const objVal = evalNode(node.object, ctx);
      return objVal?.[prop];
    }
    case 'ArrayExpr': {
      const arr = node.elements.map((el) => evalNode(el, ctx));
      const literalElements = node.elements.every((el) =>
        ['NumberLiteral', 'StringLiteral', 'BooleanLiteral', 'NullLiteral'].includes(el.type),
      );
      if (literalElements && !Object.prototype.hasOwnProperty.call(arr, '__literal')) {
        Object.defineProperty(arr, '__literal', { value: true });
      }
      return arr;
    }
    case 'ObjectExpr': {
      const out = {};
      for (const { key, value } of node.properties) {
        out[key] = evalNode(value, ctx);
      }
      return out;
    }
    case 'UnaryExpr': {
      const v = evalNode(node.argument, ctx);
      switch (node.operator) {
        case '+':
          return Array.isArray(v) ? v : +v;
        case '-':
          if (Array.isArray(v)) return v.map((n) => -n);
          return -v;
        default:
          throw new Error(`Unknown operator ${node.operator}`);
      }
    }
    case 'BinaryExpr': {
      const l = evalNode(node.left, ctx);
      const r = evalNode(node.right, ctx);
      const op = node.operator;
      if (hasFunction(l) || hasFunction(r)) {
        return (settings) => {
          const L = resolveValue(l, settings);
          const R = resolveValue(r, settings);
          switch (op) {
            case '+':
              if (Array.isArray(L) && Array.isArray(R)) return L.concat(R);
              if (Array.isArray(L) && typeof R === 'number') return L.map((n) => n + R);
              if (typeof L === 'number' && Array.isArray(R)) return R.map((n) => L + n);
              return L + R;
            case '-':
              if (Array.isArray(L) && Array.isArray(R)) return L.map((n, i) => n - R[i]);
              if (Array.isArray(L) && typeof R === 'number') return L.map((n) => n - R);
              if (typeof L === 'number' && Array.isArray(R)) return R.map((n) => L - n);
              return L - R;
            case '*':
              if (Array.isArray(L) && Array.isArray(R)) return L.map((n, i) => n * R[i]);
              if (Array.isArray(L) && typeof R === 'number') {
                const out = [];
                for (let i = 0; i < R; i++) out.push(...L);
                return out;
              }
              if (typeof L === 'number' && Array.isArray(R)) {
                const out = [];
                for (let i = 0; i < L; i++) out.push(...R);
                return out;
              }
              return L * R;
            case '/':
              if (Array.isArray(L) && typeof R === 'number') return L.map((n) => n / R);
              if (typeof L === 'number' && Array.isArray(R)) return R.map((n) => L / n);
              return L / R;
            case '<':
              return L < R;
            case '>':
              return L > R;
            default:
              throw new Error(`Unknown operator ${op}`);
          }
        };
      }
      switch (op) {
        case '+':
          if (Array.isArray(l) && Array.isArray(r)) return l.concat(r);
          if (Array.isArray(l) && typeof r === 'number') return l.map((n) => n + r);
          if (typeof l === 'number' && Array.isArray(r)) return r.map((n) => l + n);
          return l + r;
        case '-':
          if (Array.isArray(l) && Array.isArray(r)) return l.map((n, i) => n - r[i]);
          if (Array.isArray(l) && typeof r === 'number') return l.map((n) => n - r);
          if (typeof l === 'number' && Array.isArray(r)) return r.map((n) => l - n);
          return l - r;
        case '*':
          if (Array.isArray(l) && Array.isArray(r)) return l.map((n, i) => n * r[i]);
          if (Array.isArray(l) && typeof r === 'number') {
            const out = [];
            for (let i = 0; i < r; i++) out.push(...l);
            return out;
          }
          if (typeof l === 'number' && Array.isArray(r)) {
            const out = [];
            for (let i = 0; i < l; i++) out.push(...r);
            return out;
          }
          return l * r;
        case '/':
          if (Array.isArray(l) && typeof r === 'number') return l.map((n) => n / r);
          if (typeof l === 'number' && Array.isArray(r)) return r.map((n) => l / n);
          return l / r;
        case '<':
          return l < r;
        case '>':
          return l > r;
        default:
          throw new Error(`Unknown operator ${op}`);
      }
    }
    case 'TernaryExpr': {
      const test = evalNode(node.test, ctx);
      const consequent = evalNode(node.consequent, ctx);
      const alternate = evalNode(node.alternate, ctx);
      if (hasFunction(test) || hasFunction(consequent) || hasFunction(alternate)) {
        return (settings) =>
          resolveValue(test, settings)
            ? resolveValue(consequent, settings)
            : resolveValue(alternate, settings);
      }
      return test ? consequent : alternate;
    }
    case 'CallExpr':
      return evalCall(node, ctx);
    default:
      throw new Error(`Unsupported node type: ${node.type}`);
  }
}

function evalArgs(arglist, ctx) {
  if (arglist.named) {
    const params = {};
    const names = [];
    for (const [k, v] of Object.entries(arglist.named)) {
      params[k] = evalNode(v, ctx);
      names.push(k);
    }
    return { args: [params], params, paramNames: names };
  }
  const arr = arglist.positional.map((a) => evalNode(a, ctx));
  return { args: arr, params: null, paramNames: null };
}

function hasFunction(v) {
  if (typeof v === 'function') return true;
  if (Array.isArray(v)) return v.some(hasFunction);
  if (v && typeof v === 'object') return Object.values(v).some(hasFunction);
  return false;
}

function resolveValue(v, settings) {
  if (typeof v === 'function') return v(settings);
  if (Array.isArray(v)) {
    const out = v.map((x) => resolveValue(x, settings));
    if (v.__enum && !Object.prototype.hasOwnProperty.call(out, '__enum')) {
      Object.defineProperty(out, '__enum', { value: v.__enum });
    }
    if (v.__literal && !Object.prototype.hasOwnProperty.call(out, '__literal')) {
      Object.defineProperty(out, '__literal', { value: v.__literal });
    }
    return out;
  }
  if (v && typeof v === 'object') {
    const out = {};
    for (const [k, val] of Object.entries(v)) {
      out[k] = resolveValue(val, settings);
    }
    return out;
  }
  return v;
}

function evalCall(node, ctx) {
  const input = node.input ? evalNode(node.input, ctx) : undefined;
  const { args, params, paramNames } = evalArgs(node.args, ctx);
  const callee = node.callee;
  if (callee.type === 'Identifier') {
    const name = callee.name;
    const fn = ctx.operations[name];
    if (typeof fn === 'function') {
      if (fn.__thunk || hasFunction(args) || hasFunction(input)) {
        return (settings) => fn(...resolveValue(args, settings));
      }
      return fn(...args);
    }
    const out = { op: name, args, input };
    out.__effectName = name;
    if (paramNames) {
      out.__paramNames = paramNames;
      out.__params = params;
    }
    return out;
  }
  const fn = evalNode(callee, ctx);
  if (typeof fn === 'function') {
    if (fn.__thunk || hasFunction(args) || hasFunction(input)) {
      return (settings) => fn(...resolveValue(args, settings));
    }
    return fn(...args);
  }
  throw new Error('Unsupported callee');
}

export function evaluate(ast, ctx = defaultContext) {
  return evalNode(ast.body, ctx);
}

File: noisemaker/dsl/evaluator.py
from .builtins import defaultContext
import enum


def _has_function(v):
    if callable(v):
        return True
    if isinstance(v, list):
        return any(_has_function(x) for x in v)
    if isinstance(v, dict):
        return any(_has_function(x) for x in v.values())
    return False


def _resolve_value(v, settings):
    if callable(v) and not isinstance(v, enum.EnumMeta):
        return v(settings)
    if isinstance(v, list):
        return [_resolve_value(x, settings) for x in v]
    if isinstance(v, dict):
        return {k: _resolve_value(x, settings) for k, x in v.items()}
    return v


def _apply_unary(val, op):
    if op == '+':
        return val if isinstance(val, list) else +val
    if op == '-':
        if isinstance(val, list):
            return [-x for x in val]
        return -val
    raise ValueError(f"Unknown operator {op}")


def _apply_binary(l, r, op):
    if op == '+':
        if isinstance(l, list) and isinstance(r, list):
            return l + r
        if isinstance(l, list) and isinstance(r, (int, float)):
            return [n + r for n in l]
        if isinstance(l, (int, float)) and isinstance(r, list):
            return [l + n for n in r]
        return l + r
    if op == '-':
        if isinstance(l, list) and isinstance(r, list):
            return [n - r[i] for i, n in enumerate(l)]
        if isinstance(l, list) and isinstance(r, (int, float)):
            return [n - r for n in l]
        if isinstance(l, (int, float)) and isinstance(r, list):
            return [l - n for n in r]
        return l - r
    if op == '*':
        if isinstance(l, list) and isinstance(r, list):
            return [n * r[i] for i, n in enumerate(l)]
        if isinstance(l, list) and isinstance(r, (int, float)):
            out = []
            for _ in range(int(r)):
                out.extend(l)
            return out
        if isinstance(l, (int, float)) and isinstance(r, list):
            out = []
            for _ in range(int(l)):
                out.extend(r)
            return out
        return l * r
    if op == '/':
        if isinstance(l, list) and isinstance(r, (int, float)):
            return [n / r for n in l]
        if isinstance(l, (int, float)) and isinstance(r, list):
            return [l / n for n in r]
        return l / r
    if op == '<':
        return l < r
    if op == '>':
        return l > r
    raise ValueError(f"Unknown operator {op}")

def _eval_node(node, ctx):
    t = node['type']
    if t == 'NumberLiteral':
        return node['value']
    if t == 'StringLiteral':
        return node['value']
    if t == 'BooleanLiteral':
        return node['value']
    if t == 'NullLiteral':
        return None
    if t == 'Identifier':
        operations = ctx.get('operations', {})
        surfaces = ctx.get('surfaces', {})
        enums = ctx.get('enums')
        name = node['name']
        if name in operations:
            return operations[name]
        if name in surfaces:
            return surfaces[name]
        if enums and hasattr(enums, name):
            return getattr(enums, name)
        return name
    if t == 'MemberExpr':
        obj = node['object']
        prop = node['property']['name']
        if obj['type'] == 'Identifier' and ctx.get('enumMethods'):
            methods = ctx['enumMethods'].get(obj['name'])
            if methods and prop in methods:
                return methods[prop]
        obj_val = _eval_node(obj, ctx)
        if isinstance(obj_val, dict):
            return obj_val.get(prop)
        return getattr(obj_val, prop, None)
    if t == 'ArrayExpr':
        return [_eval_node(el, ctx) for el in node['elements']]
    if t == 'ObjectExpr':
        out = {}
        for pair in node['properties']:
            out[pair['key']] = _eval_node(pair['value'], ctx)
        return out
    if t == 'UnaryExpr':
        val = _eval_node(node['argument'], ctx)
        op = node['operator']
        if _has_function(val):
            return lambda settings: _apply_unary(_resolve_value(val, settings), op)
        return _apply_unary(val, op)
    if t == 'BinaryExpr':
        l = _eval_node(node['left'], ctx)
        r = _eval_node(node['right'], ctx)
        op = node['operator']
        if _has_function(l) or _has_function(r):
            return lambda settings: _apply_binary(_resolve_value(l, settings), _resolve_value(r, settings), op)
        return _apply_binary(l, r, op)
    if t == 'TernaryExpr':
        test = _eval_node(node['test'], ctx)
        consequent = _eval_node(node['consequent'], ctx)
        alternate = _eval_node(node['alternate'], ctx)
        if _has_function(test) or _has_function(consequent) or _has_function(alternate):
            return lambda settings: (
                _resolve_value(consequent, settings)
                if _resolve_value(test, settings)
                else _resolve_value(alternate, settings)
            )
        return _eval_node(node['consequent'], ctx) if test else _eval_node(node['alternate'], ctx)
    if t == 'CallExpr':
        return _eval_call(node, ctx)
    raise ValueError(f"Unsupported node type: {t}")

def _eval_args(arglist, ctx):
    if 'named' in arglist:
        params = {}
        names = []
        for k, v in arglist['named'].items():
            params[k] = _eval_node(v, ctx)
            names.append(k)
        return {'args': [params], 'params': params, 'paramNames': names}
    arr = [_eval_node(a, ctx) for a in arglist.get('positional', [])]
    return {'args': arr, 'params': None, 'paramNames': None}

def _eval_call(node, ctx):
    input_val = _eval_node(node['input'], ctx) if node.get('input') else None
    evaled = _eval_args(node['args'], ctx)
    args = evaled['args']
    params = evaled['params']
    param_names = evaled['paramNames']
    callee = node['callee']
    if callee['type'] == 'Identifier':
        name = callee['name']
        fn = ctx.get('operations', {}).get(name)
        if callable(fn):
            if getattr(fn, "__thunk", False) or _has_function(args) or _has_function(input_val):
                return lambda settings: fn(*_resolve_value(args, settings))
            return fn(*args)
        out = {'op': name, 'args': args, 'input': input_val, '__effectName': name}
        if param_names:
            out['__paramNames'] = param_names
            out['__params'] = params
        return out
    fn = _eval_node(callee, ctx)
    if callable(fn):
        if getattr(fn, "__thunk", False) or _has_function(args) or _has_function(input_val):
            return lambda settings: fn(*_resolve_value(args, settings))
        return fn(*args)
    raise ValueError('Unsupported callee')

def evaluate(ast, ctx=defaultContext):
    return _eval_node(ast['body'], ctx)
